### ISSUE TITLE
Wrong initialization order between clustered mode or not

### ISSUE THREAD

**Issue by clems159** [gentics/mesh#344]
https://github.com/gentics/mesh/blob/23e8e5a7520cb955340477ac719dbcbb621e3a81/core/src/main/java/com/gentics/mesh/cli/BootstrapInitializerImpl.java#L287

Hello @Jotschi 
you are not doing the same thing if clustered or not,

in cluster mode you are starting db server first and you load the /mesh/config/orientdb-server-config.xml

in non cluster mode, it seems that you start server after using it, so i assume it is loaded with default orientdb-server-config.xml properties.

in cluster mode we can see this log
```
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    | 10:26:16.122 [mesh-master] DEBUG [main] - Found database service provider {class com.gentics.mesh.graphdb.OrientDBDatabase}
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    | 10:26:16.438 [mesh-master] INFO  [main] - Reflections took 236 ms to scan 1 urls, producing 136 keys and 342 values
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    | 10:26:23.617 [mesh-master] INFO  [main] - Extracting OrientDB Studio
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    | 10:26:25.246 [mesh-master] INFO  [main] - Starting OrientDB Server
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    | 10:26:25.766 [mesh-master] DEBUG [main] - Effective orientdb server configuration:<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    | <orient-server>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |     <handlers>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |         <handler class="com.orientechnologies.orient.graph.handler.OGraphServerHandler">
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |             <parameters>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                 <parameter name="enabled" value="true"/>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                 <parameter name="graph.pool.max" value="50"/>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |             </parameters>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |         </handler>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |         <handler class="com.orientechnologies.orient.server.hazelcast.OHazelcastPlugin">
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |             <parameters>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                 <parameter name="nodeName" value="mesh-master@0-17-2"/>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                 <parameter name="enabled" value="true"/>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                 <parameter name="configuration.db.default"
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                            value="${ORIENTDB_HOME}/config/default-distributed-db-config.json"/>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                 <parameter name="configuration.hazelcast" value="${ORIENTDB_HOME}/config/hazelcast.xml"/>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |             </parameters>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |         </handler>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |         <handler class="com.orientechnologies.orient.server.handler.OJMXPlugin">
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |             <parameters>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                 <parameter name="enabled" value="true"/>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                 <parameter name="profilerManaged" value="true"/>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |             </parameters>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |         </handler>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |         <handler class="com.orientechnologies.orient.server.handler.OAutomaticBackup">
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |             <parameters>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                 <parameter name="enabled" value="true"/>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                 <parameter name="config" value="${ORIENTDB_HOME}/config/automatic-backup.json"/>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |             </parameters>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |         </handler>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |     </handlers>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |     <network>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |         <protocols>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |             <protocol name="binary"
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                       implementation="com.orientechnologies.orient.server.network.protocol.binary.ONetworkProtocolBinary"/>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |             <protocol name="http"
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                       implementation="com.orientechnologies.orient.server.network.protocol.http.ONetworkProtocolHttpDb"/>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |         </protocols>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |         <listeners>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |             <listener protocol="binary" ip-address="0.0.0.0" port-range="2500-2430" socket="default"/>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |             <listener protocol="http" ip-address="0.0.0.0" port-range="2480-2490" socket="default">
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                 <parameters>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                     <!-- Connection's custom parameters. If not specified the global configuration will be taken -->
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                     <parameter name="network.http.charset" value="utf-8"/>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                     <parameter value="true" name="network.http.jsonResponseError"/>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                 </parameters>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                 <commands>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                     <command
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                             pattern="GET|www GET|studio/ GET| GET|*.htm GET|*.html GET|*.xml GET|*.jpeg GET|*.jpg GET|*.png GET|*.gif GET|*.js GET|*.css GET|*.swf GET|*.ico GET|*.txt GET|*.otf GET|*.pjs GET|*.svg GET|*.json GET|*.woff GET|*.woff2 GET|*.ttf GET|*.svgz"
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                             implementation="com.orientechnologies.orient.server.network.protocol.http.command.get.OServerCommandGetStaticContent">
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                         <parameters>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                             <!-- Don't cache html resources in development mode -->
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                             <entry name="http.cache:*.htm *.html"
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                                    value="Cache-Control: no-cache, no-store, max-age=0, must-revalidate\r\nPragma: no-cache"/>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                             <!-- Default caching -->
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                             <entry name="http.cache:default" value="Cache-Control: max-age=120"/>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                         </parameters>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                     </command>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                     <command pattern="GET|gephi/*"
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                              implementation="com.orientechnologies.orient.graph.server.command.OServerCommandGetGephi"/>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |                 </commands>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |             </listener>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |         </listeners>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |         <cluster>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |         </cluster>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |     </network>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |     <users>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |         <user name="root" password="finger" resources="*"/>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |         <user name="admin" password="finger" resources="*"/>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |     </users>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |     <properties>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |         <!-- DATABASE POOL: size min/max -->
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |         <entry name="db.pool.min" value="1"/>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |         <entry name="db.pool.max" value="50"/>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |         <!-- PROFILER: configures the profiler as <seconds-for-snapshot>,<archive-snapshot-size>,<summary-size> -->
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |         <entry name="profiler.enabled" value="false"/>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |         <entry name="plugin.directory" value="/mesh/orientdb-plugins"/>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |         <entry name="server.database.path" value="/mesh/data/graphdb"/>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |         <!-- LOG: enable/Disable logging. Levels are: finer, fine, finest, info, warning -->
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |         <entry name="log.console.level" value="finest"/>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |         <entry name="log.file.level" value="finest"/>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |     </properties>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    | </orient-server>
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |
cms_mesh-master.1.nv0o79rbpzce@linuxkit-025000000001    |
```

in non cluster mode we never see this output.

**Comment by Jotschi** [comment]
@clems159 In non cluster mode the OrientDB server will not be started (default setting). The OrientDB server is just needed for clustering and to access the embedded OrientDB studio. You can however also enable the OrientDB Server startup in non cluster mode by setting the `storage.startServer` setting in the `mesh.yml` to true.
See https://getmesh.io/docs/beta/administration-guide.html#_storage_options

I would however only enable the server in non clustered mode if you need to debug the database.
The OrientDB server will automatically be started if clustering is *enabled* even if the `storage.startServer` since it is required for clustering.

Hope this helps.

**Comment by cschockaert** [comment]
ok thanks, i thought that to use orientdb server startup was needed. but it is not (in non cluster mode)

actually my problem was to activate replication with single node mode on (cluster off), i'll try with storage.startServer=true settings.

in fact, i'm working on getting getmesh scalable and HA, but in master / master ive got some synhcronization problems database corruption etc in certains circunstances (high usage node update/removal etc...)

i switched to 1 master / x replica typo, using the role replica from orientdb 2.33.X, It's working OOTB BUT when i restart a replica instance, some time it fall in a loop with the master,

-> retreive backup from master, switch to unaivalable, but saying that he's db is newer than master one. infinite loop between master and replicas.

after that, i switched back for now to 1 single getmesh instance highly backuped (i tested multiple backup restore scenario successfully but only in single mode scenario).

i'm using native AutomaticBackup plugin from orientdb to acheive that.

i had some problem with the getmesh backup/restore endpoint.

**Comment by Jotschi** [comment]
@cschockaert Let me try to give you some feeback.

> activate replication with single node mode on (cluster off), 
I assume you want run one master and multiple "read-only" slaves. This is currently not supported. We have plans to add this in the future but add the moment Gentics Mesh does not support this. (Although OrientDB is capable of running in this mode)

> infinite loop between master and replicas.
Did it never stop to replicate? If this is so than I can tell you that I have already reported that issue and it has been fixed by the OrientDB team. I'll release a new Mesh version once 2.2.34 has been released. 

> i had some problem with the getmesh backup/restore endpoint.
Backup works in our tests but you are required to restart mesh once you restored the backup.

**Comment by cschockaert** [comment]
ok, i saw you opened an issue about making getmesh readonly mode viable.

my tests was the followings: instanciate a master and some replicas with docker swarm, using an elasticsearch cluster too.

working directly on master node, add/remove/update nodes.
everythings is fine.

going on mesh-ui replica instance:

doing some search, read node: OK
change node : KO but its normal db is in read only mode.
all is working OOTB. because it's seems that getmesh can work even if it cant write things on his associated db. (not tested all cases through)..

but after some restart of replicas i got this inifnite loop you reported to orientdb team.

here my conf:
```
version: '3.5'

# directory structure :https://getmesh.io/docs/beta/administration-guide.html#_backup_recovery

#data
#    binaryFiles
#    binaryImageCache
#    graphdb
#    tmp
# conf
#    keystore.jceks
#    default-distributed-db-config.json
#    orientdb-server-config.xml
#    security.json
#    hazelcast.xml
#    mesh-ui-config.js
#    mesh.yml
#elasticsearch


services:
  mesh-master:
    image: registry/getmesh:0.17.3
    networks:
      default:
      traefik-net:
    ports:
      - 8080
      - 2480
    environment:
      ORIENTDB_BACKUP: "true"
      MESH_NODE_NAME: "mesh-master"
      MESH_CLUSTER_INIT: "false"
      MESH_CLUSTER_ENABLED: "true" #in first start set MESH_CLUSTER_ENABLED to false, then reset it to true
      MESH_CLUSTER_NAME: "mesh.cluster"
      MESH_CLUSTER_NETWORK_HOST: "0.0.0.0"
      MESH_ELASTICSEARCH_START_EMBEDDED: "false"
      MESH_ELASTICSEARCH_URL: "http://elasticsearch:9200/"
    command: java -jar mesh.jar
    volumes:
      - getmesh_binaries:/binaryFiles  #shared directory between nodes
      - getmesh_config:/config
      - getmesh_master_data:/data
    deploy:
      replicas: 1
      labels:
        - traefik.frontend.rule=Host:getmesh
        - traefik.port=8080
        - traefik.docker.network=traefik-net
  mesh-replica_1:
      image: registry/getmesh:0.17.3
      networks:
        default:
      ports:
        - 8080
        - 2480
      environment:
        ORIENTDB_BACKUP: "false"
        MESH_GRAPH_DB_DIRECTORY: "data/{{.Service.Name}}/graphdb"
        MESH_NODE_NAME: "{{.Service.Name}}"
      volumes:
        - getmesh_binaries:/binaryFiles  #shared directory between nodes
        - getmesh_config:/config
        - getmesh_master_data:/data
      command: java -jar mesh.jar
      deploy:
        replicas: 0
        #endpoint_mode: vip
  mesh-replica_2:
      image: registry/getmesh:0.17.3
      networks:
        default:
      ports:
        - 8080
        - 2480
      environment:
        ORIENTDB_BACKUP: "false"
        MESH_GRAPH_DB_DIRECTORY: "data/{{.Service.Name}}/graphdb"
        MESH_NODE_NAME: "{{.Service.Name}}"
      volumes:
        - getmesh_binaries:/binaryFiles  #shared directory between nodes
        - getmesh_config:/config
      command: java -jar mesh.jar
      deploy:
        replicas: 0
        #endpoint_mode: vip

  elasticsearch:
    image: docker.elastic.co/elasticsearch/elasticsearch:6.1.3
    ports:
     - 9200
    environment:
      ES_JAVA_OPTS: "-Xms512m -Xmx512m"
      cluster.name: "docker-cluster"
      node.name: "es_{{.Task.Slot}}-{{.Service.Name}}-{{.Service.ID}}"
      path.data: "/usr/share/elasticsearch/data/es_{{.Task.Slot}}-{{.Service.Name}}"
      #bootstrap.memory_lock: "true"
      http.cors.allow-origin: "*"
      http.cors.enabled: "true"
      http.cors.allow-headers: "X-Requested-With,X-Auth-Token,Content-Type,Content-Length,Authorization"
      http.cors.allow-credentials: "true"
      discovery.zen.ping.unicast.hosts: "tasks.elasticsearch" #used to query against direct ip name (like in dnsrr mode)
      discovery.zen.minimum_master_nodes: "2"
      network.host: _eth0:ipv4_ # force use another interface to not announce vip (which is in loopback itf)
    volumes:
      - elasticsearch_data:/usr/share/elasticsearch/data
    ulimits:
      memlock:
        soft: -1
        hard: -1
    deploy:
      replicas: 2
      endpoint_mode: vip
    networks:
      default:

networks:
  default:
  traefik-net:
    external: true

volumes:
  getmesh_config:
   driver: hjdr4plugins/docker-volume-glusterfs
#    driver: local
  getmesh_binaries:
    driver: hjdr4plugins/docker-volume-glusterfs
#    driver: local
  getmesh_master_data:
    driver: hjdr4plugins/docker-volume-glusterfs
#    driver: local
  elasticsearch_data:
    driver: hjdr4plugins/docker-volume-glusterfs
#    driver: local
```

**Comment by cschockaert** [comment]
hazelcast to get it work

```
<network>
        <port auto-increment="true" port-count="100">5701</port>
        <join>
            <multicast enabled="false">
                <!--<multicast-group>235.1.1.1</multicast-group>-->
                <!--<multicast-port>2434</multicast-port>-->
            </multicast>
            <tcp-ip enabled="true">
                <member>mesh-master</member> <!-- 1 master -->

                <!-- x replicas disabled for now because UNDER TESTING -->
                <!--<member>mesh-replica_1</member>-->
                <!--<member>mesh-replica_2</member>-->
            </tcp-ip>
        </join>
        <!--<interfaces enabled="true">-->
        <!--<interface>10.0.0.*</interface>-->
        <!--</interfaces>-->
    </network>
```

**Comment by Jotschi** [comment]
> but after some restart of replicas i got this inifnite loop you reported to orientdb team.

You should be able to mitigate this by deleting the graphdb folder. I know this sounds odd but this will force a full sync of the db without a differential sync being considered and thus the issue will be avoided.

**Comment by cschockaert** [comment]
yes, it worked, but cannot use that in production because, when master is backuping he's not available

**Comment by cschockaert** [comment]
well, will be waiting next update :) thanks

**Comment by Jotschi** [comment]
@cschockaert We also hope to offer a enterprise version of Gentics Mesh soon which will support live differential backup.

**Comment by cschockaert** [comment]
for now, we'll got with this one:

Non-Blocking Backup

backup.sh script supports non-blocking backup if the OS supports LVM). Example:

./backup.sh plocal:../database/testdb admin admin /dest/folder/backup.zip lvm

we are using lvm behind our glusterfs bricks

**Comment by Jotschi** [comment]
@cschockaert Sounds good. Keep in mind that Elasticsearch does not like NFS. The need for open file handles is crazy. Not sure how it deals with GlusterFS. OrientDB had less issues with NFS. (I did not test glusterfs).

**Comment by cschockaert** [comment]
ok, well, we have to do some benchmarks, but i thinks that for elasticsearch volume are not required. in fact we just have to scale a lot of service with a lot of shard replications, if one fail swarm will popup new instance.

it's just that if all failed and data is lost we should perform a reindex from getmesh, 
index process seems to be reliable and working fine and fast.

i will look into another volume driver for db storage.

**Comment by Jotschi** [comment]
@cschockaert Reindex can take some time on larger systems. We also plan to add a way to trigger a differential reindex which will be much faster.

**Comment by cschockaert** [comment]
ok thanks, ill try to only setup binaries on gluster but put other graphdb and elasticsearch on a classic disk storage.

**Comment by cschockaert** [comment]
talking of that, you should create a dedicated volume in your docker file for binaryFiles since it must not be shared with /data volume

**Comment by Jotschi** [comment]
@cschockaert That should be fairly trivial. I'll discuss that with the team. Thanks for the feedback.

We also plan to add S3 binary storage support for Gentics Mesh. That way https://www.minio.io/ could be used to store all binaries.

**Comment by Jotschi** [comment]
@cschockaert We have reworked the volumes in the docker images. Changes have been released with 0.27.0 - I have also updated the docker compose clustering example. It now contains an example on how to add a dedicated mesh node which handles the backup process.

https://github.com/gentics/mesh-compose/tree/clustering
